---
title: "Human Genome Analysis Lab 11: Analysis of transcriptome (RNA-Seq) data"
output:
  html_document:
    toc: true
    toc_depth: 4
    theme: united
    highlight: tango
editor_options: 
  chunk_output_type: console
---

## Learning Objectives

* Understand the basic workflow for RNAseq
* Data normalization using DESeq
* Data exploration using PCA, PCO and MDS

## Background

When you are trying to understand a new analysis method, a good strategy is to 

1. Read background material so that you conceptual understand the work flow.
2. Go through a tutorial and/or R vignette with the tools/packages you will need.
3. Apply the tutorials/vignettes to your experiment.

Today we will go through steps 1 and 2. For an introduction to RNAseq analysis we will discuss this [Galaxy workflow](https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/rb-rnaseq/tutorial.html). The steps involving __read mapping__, __transcript quantification__, and __read counting__ are usually done on a Unix computer using the command line or an interface to a Unix computer such as Galaxy. We will go through these steps in the tutorial and then in a later lab on a Unix virtual machine or the MGHPCC. Today in lab we will work on __differential expression analysis in R__. 


## Introduction to a RNA-Seq differential expression workflow

This lab will go through __differential expression analysis in R__ using DESeq2 along with other Bioconductor and core R packages. 

The data used in this workflow is an RNA-Seq experiment of airway smooth muscle cells treated with dexamethasone, a synthetic glucocorticoid steroid with anti-inflammatory effects. Glucocorticoids are used, for example, in asthma patients to prevent or reduce inflammation of the airways. In the experiment, four primary human airway smooth muscle cell lines were treated with 1 micromolar dexamethasone for 18 hours. For each of the four cell lines, we have a treated and an untreated sample. The reference for the experiment is:

Himes BE, Jiang X, Wagner P, Hu R, Wang Q, Klanderman B, Whitaker RM, Duan Q, Lasky-Su J, Nikolos C, Jester W, Johnson M, Panettieri R Jr, Tantisira KG, Weiss ST, Lu Q. "RNA-Seq Transcriptome Profiling Identifies CRISPLD2 as a Glucocorticoid Responsive Gene that Modulates Cytokine Function in Airway Smooth Muscle Cells." PLoS One. 2014 Jun 13;9(6):e99625. PMID: 24926665. GEO: GSE52778.

To install the packages and data needed to complete this tutorial on your computer (you do not need to do this on Posit). This assumes your have already installed the core Bioconductor (see Lab 9).

```{r, eval=FALSE}
# you do not need to do this on Posit
# Unity users should allocate 12 GB RAM to the JupyterHub server and set time to 4 hrs
BiocManager::install("rnaseqGene")
```

Unity users may get the following message

```{r}
# Installation paths not writeable, unable to update packages
#   path: /modules/apps/miniconda/4.8.3/envs/jupyterhub-stable/lib/R/library
#   packages:
#     blob, broom, bslib, callr, caret, cli, cluster, commonmark, cpp11, crayon, crul, curl, data.table, DBI, dbplyr, desc, digest, dplyr, dtplyr, e1071, evaluate, farver,
#     fontawesome, forcats, foreign, future, future.apply, gargle, generics, ggplot2, glmnet, globals, googlesheets4, gtable, hardhat, haven, hms, htmltools, httpuv, httr,
#     ipred, isoband, jsonlite, later, lava, lifecycle, lobstr, maps, MASS, Matrix, mgcv, modelr, nlme, nnet, openssl, parallelly, pbdZMQ, pillar, pkgload, progressr, proxy,
#     purrr, quantmod, randomForest, Rcpp, readr, readxl, recipes, reprex, rlang, rmarkdown, rpart, rstudioapi, rvest, sass, scales, shiny, stringi, survival, sys, testthat,
#     tibble, tidyr, tidyselect, tidyverse, timeDate, tinytex, uuid, vctrs, viridisLite, vroom, xfun, xts, yaml, zoo
# Old packages: 'ade4', 'BiocManager', 'brew', 'chron', 'devtools', 'downlit', 'DT', 'duckdb', 'gert', 'gh', 'gitcreds', 'igraph', 'knitr', 'latticeExtra', 'mapdata',
#   'markdown', 'matrixStats', 'neonUtilities', 'openxlsx', 'processx', 'ps', 'qdapRegex', 'R.methodsS3', 'R.oo', 'R.utils', 'RCurl', 'roxygen2', 'rversions', 'sp',
 #  'stringdist', 'stringr', 'tm', 'usethis', 'vegan', 'XML', 'zip'
# Update all/some/none? [a/s/n]: 
```

select a

We will go step wise through a Bioconductor workflow vignette [RNA-seq workflow: gene-level exploratory analysis and differential expression](https://master.bioconductor.org/packages/release/workflows/html/rnaseqGene.html). We will start by opening the html file associated with the workflow and downloading the R script to our computer. 

__IMPORTANT__
When running the following the first time
```{r, eval=FALSE}
library("tximeta")
se <- tximeta(coldata)
```

You may get the following message

```{r}
# tximeta needs a BiocFileCache directory to access and save TxDb objects.
# Do you wish to use the default directory: '/cloud/home/r555412/.cache/R/BiocFileCache'?
# If not, a temporary directory that is specific to this R session will be used.
# You can always change this directory later by running: setTximetaBFC()
# Or enter [0] to exit and set this directory manually now.
# This location can also be set by environmental variable TXIMETA_HUB_CACHE. 

# 1: Yes (use default)
# 2: No (use temp)
```

Select 1 (Yes)

Then then a message similar to this may appear
```{r}
# /cloud/home/r555412/.cache/R/BiocFileCache
#   does not exist, create directory? (yes/no): 
```

Select yes


## Exercises

Work through the [DESeq2 vignette](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html). I have worked through the r code, installed the relevant libraries and included it in the files with the lab on Posit. Here is a direct link to the [DESeq2 R script](https://bioconductor.org/packages/release/bioc/html/DESeq2.html0. There are some spots in the html version that are commented out in the R script. This may not work and you are not responsible for them. For the lab convert the html file to a Rmd file as you have down with the other lab examples. Turn in the knitted html file.


Here is another option in greater detail - https://mperalc.gitlab.io/bulk_RNA-seq_workshop_2021/qc.html
